codex/merkletree/merkletree.nim

```diff
@@ -0,0 +1,111 @@
+import std/sequtils
+import std/math
+import std/bitops
+import std/sugar
+
+import pkg/libp2p
+import pkg/stew/byteutils
+
+import pkg/questionable
+import pkg/questionable/results
+
+
+type
+  MerkleHash* = MultiHash
+  MerkleTree* = ref object of RootObj
+    nodes: seq[seq[MerkleHash]]
+  MerkleTreeBuilder* = ref object of RootObj
+    mcodec: ?MultiCodec
+    leaves: seq[MerkleHash]
+  MerkleProof* = ref object of RootObj
+    path: seq[MerkleHash]
+
+proc `$`*(t: MerkleTree): string =
+  result &= "height: " & $t.nodes.len
+  result &= "\nnodes: " & $t.nodes
+
+proc addLeaf*(self: MerkleTreeBuilder, hash: MerkleHash): ?!void =
+  if codec =? self.mcodec:
+    if codec != hash.mcodec:
+      return failure("Expected codec is " & $codec & " but " & $hash.mcodec & " received")
+  else:
+    self.mcodec = hash.mcodec.some
+
+  self.leaves.add(hash)
+  return success()
+
+func calcTreeHeight(leavesCount: int): int =
+  if isPowerOfTwo(leavesCount): 
+    fastLog2(leavesCount) + 1
+  else:
+    fastLog2(leavesCount) + 2
+
+proc build*(self: MerkleTreeBuilder): ?!MerkleTree =
+  let height = calcTreeHeight(self.leaves.len)
+  var nodes = newSeq[seq[MerkleHash]](height)
+
+  without mcodec =? self.mcodec and 
+          digestSize =? self.leaves.?[0].?size:
+    return failure("Unable to determine codec, no leaves were added")
+
+  var buf = newSeq[byte](digestSize * 2)
+  proc combine(l, r: MerkleHash): ?!MerkleHash =
+    copyMem(addr buf[0], unsafeAddr l.data.buffer[0], digestSize)
+    copyMem(addr buf[digestSize], unsafeAddr r.data.buffer[0], digestSize)
+
+    MultiHash.digest($mcodec, buf).mapErr(
+      c => newException(CatchableError, "Error calculating hash using codec " & $mcodec & ": " & $c)
+    )
+
+  # copy leaves
+  nodes[0] = newSeq[MerkleHash](self.leaves.len)
+  for j in 0..<self.leaves.len:
+    nodes[0][j] = self.leaves[j]
+
+  # calculate internal nodes
+  for i in 1..<height:
+    let levelWidth = (nodes[i-1].len + 1) div 2
+    nodes[i] = newSeq[MerkleHash](levelWidth)
```

Comment:
Since this is a balanced merkle tree could we just use a flat `seq[MerkleHash]` and do indexing children via powers of 2? It's a bit more work here, but a flat `seq` is easy to serialize and copy. 

I guess it'd make add dynamic leaves a bit more expensive though?

Reply:
It's not necessarily a balanced tree, since not all subtrees are balanced. We could still make it with a 1D array but it makes indexing a bit more complicated though. For now I would stay with 2D array and change it to 1D once it necessary (probably for serialization).

Reply:
Yes, I agree, as we discussed in the initial call, this should be a flat array for the reasons that @elcritch points out.

Reply:
Ok, switched to 1D array.